File: OJ/Leetcode/Algorithm/402. Remove K Digits.java
class Solution {
    public String removeKdigits(String num, int k) {
        int len = num.length();
        if(k == len)    return "0";
        Stack<Character> st = new Stack<>();
        for(int i = 0; i < len; i++)
        {
            while(k > 0 && !st.isEmpty() && st.peek() > num.charAt(i))
            {
                st.pop();
                k--;
            }
            st.push(num.charAt(i));
        }
        while(k > 0)
        {
            st.pop();
            k--;
        }

        StringBuilder sb = new StringBuilder();
        while(!st.isEmpty())
            sb.append(st.pop());
        sb.reverse();

        while(sb.length()>1 && sb.charAt(0)=='0')
            sb.deleteCharAt(0);
        return sb.toString();
    }
}

File: OJ/Leetcode/Algorithm/402. Remove K Digits.cpp
class Solution {
public:
    string removeKdigits(string num, int k) {
        if(k == num.length())   return "0";
        stack<char> st;
        for(auto ch : num)
        {
            while(k > 0 && !st.empty() && st.top() > ch)
            {
                st.pop();
                k--;
            }
            st.push(ch);
        }
        while(k-- > 0)
            st.pop();
        string str = "";
        while(!st.empty())
        {
            str += string(1, st.top());
            st.pop();
        }
        bool flag = false;
        for(int i = str.length()-1; i >= 0; i--)
        {
            if(str[i] != '0')
            {
                str = str.substr(0, i+1);
                flag = true;
                break;
            }
        }
        if(!flag)
            return "0";
        reverse(str.begin(), str.end());
        return str;
    }
};
